File: src/types/equipment/Item.ts
/**
 * @main Item
 */

import { validateSchemaCreator } from "../../validation/schema.js"
import { Errata } from "../source/_Erratum.js"
import { PublicationRefs } from "../source/_PublicationRef.js"
import { ArmorType, Encumbrance, HasAdditionalPenalties, HitZone, Protection } from "./_Armor.js"
import { MeleeWeapon, RangedWeapon, RestrictedToCultures, SanctifiedBy, StructurePoints } from "./_Weapon.js"

/**
 * @title Item
 */
export type Item = {
  /**
   * The item's identifier. An unique, increasing integer.
   * @integer
   * @minimum 1
   */
  id: number

  /**
   * The cost in silverthalers.
   */
  cost: Cost

  /**
   * The weight in kg.
   */
  weight?: Weight

  /**
   * The complexity of crafting the item.
   */
  complexity: Complexity

  type:
    | {
      tag: "Weapon"

      /**
       * @minItems 1
       */
      uses: (
        | MeleeWeapon
        | RangedWeapon
      )[]

      sanctified_by?: SanctifiedBy

      restricted_to_cultures?: RestrictedToCultures

      secondary?: {
        type: Type

        structure_points: StructurePoints
      }

      /**
       * All translations for the entry, identified by IETF language tag
       * (BCP47).
       * @minProperties 1
       */
      translations: {
        /**
         * @patternProperties ^[a-z]{2}-[A-Z]{2}$
         */
        [localeId: string]: {
          name: Name

          secondary_name?: SecondaryName

          note?: Note

          rules?: Rules

          advantage?: Advantage

          disadvantage?: Disadvantage

          errata?: Errata
        }
      }
    }
    | {
      tag: "Armor"

      protection: Protection

      encumbrance: Encumbrance

      has_additional_penalties: HasAdditionalPenalties

      armor_type: ArmorType

      hit_zone?: HitZone

      secondary?: {
        type: Type

        structure_points: StructurePoints
      }

      /**
       * All translations for the entry, identified by IETF language tag
       * (BCP47).
       * @minProperties 1
       */
      translations: {
        /**
         * @patternProperties ^[a-z]{2}-[A-Z]{2}$
         */
        [localeId: string]: {
          name: Name

          secondary_name?: SecondaryName

          note?: Note

          rules?: Rules

          advantage?: Advantage

          disadvantage?: Disadvantage

          errata?: Errata
        }
      }
    }
    | {
      tag: "Default"

      type: Type

      combat_uses:
        | {
          tag: "ImprovisedWeapon"

          /**
           * @minItems 1
           */
          uses: (
            | MeleeWeapon
            | RangedWeapon
          )[]

          sanctified_by?: SanctifiedBy

          restricted_to_cultures?: RestrictedToCultures

          /**
           * All translations for the entry, identified by IETF language tag
           * (BCP47).
           * @minProperties 1
           */
          translations: {
            /**
             * @patternProperties ^[a-z]{2}-[A-Z]{2}$
             */
            [localeId: string]: {
              advantage: Advantage

              disadvantage: Disadvantage
            }
          }
        }
        | {
          tag: "Armor"

          protection: Protection

          encumbrance: Encumbrance

          has_additional_penalties: HasAdditionalPenalties

          armor_type: ArmorType

          hit_zone?: HitZone

          /**
           * All translations for the entry, identified by IETF language tag
           * (BCP47).
           * @minProperties 1
           */
          translations: {
            /**
             * @patternProperties ^[a-z]{2}-[A-Z]{2}$
             */
            [localeId: string]: {
              advantage: Advantage

              disadvantage: Disadvantage
            }
          }
        }

      structure_points: StructurePoints

      /**
       * All translations for the entry, identified by IETF language tag
       * (BCP47).
       * @minProperties 1
       */
      translations: {
        /**
         * @patternProperties ^[a-z]{2}-[A-Z]{2}$
         */
        [localeId: string]: {
          name: Name

          secondary_name?: SecondaryName

          note?: Note

          rules?: Rules

          errata?: Errata
        }
      }
    }

  src: PublicationRefs
}

/**
 * The name of the item.
 * @minLength 1
 */
type Name = string

/**
 * An auxiliary name or label of the item, if available.
 * @minLength 1
 */
type SecondaryName = string

/**
 * The cost in silverthalers.
 */
type Cost =
  | { tag: "Free" }
  | { tag: "Various" }
  | {
    tag: "Fixed"

    /**
     * The cost in silverthalers.
     * @exclusiveMinimum 0
     */
    value: number

    /**
     * All translations for the entry, identified by IETF language tag
     * (BCP47).
     * @minProperties 1
     */
    translations?: {
      /**
       * @patternProperties ^[a-z]{2}-[A-Z]{2}$
       */
      [localeId: string]: {
        /**
         * The cost get wrapped by this text using a placeholder in the text.
         * @minLength 1
         * @pattern \\{0\\}
         */
        wrap_in_text: string
      }
    }
  }
  | {
    tag: "Range"

    /**
     * The lower bound of the cost in silverthalers.
     * @exclusiveMinimum 0
     */
    from: number

    /**
     * The upper bound of the cost in silverthalers.
     * @exclusiveMinimum 0
     */
    to: number

    /**
     * All translations for the entry, identified by IETF language tag
     * (BCP47).
     * @minProperties 1
     */
    translations?: {
      /**
       * @patternProperties ^[a-z]{2}-[A-Z]{2}$
       */
      [localeId: string]: {
        /**
         * The cost range gets wrapped by this text using a placeholder in the
         * text.
         * @minLength 1
         * @pattern \\{0\\}
         */
        wrap_in_text: string
      }
    }
  }

/**
 * The weight in kg.
 * @exclusiveMinimum 0
 */
type Weight = number

/**
 * The complexity of crafting the item.
 */
type Complexity =
  | { tag: "Primitive" }
  | { tag: "Simple" }
  | {
    tag: "Complex"

    /**
     * The AP value for the trade secret.
     * @integer
     * @minimum 1
     */
    ap_value: number
  }

enum Type {
  Ammunition = "Ammunition",
  WeaponAccessories = "WeaponAccessories",
  Clothes = "Clothes",
  TravelGearAndTools = "TravelGearAndTools",
  Illumination = "Illumination",
  BandagesAndRemedies = "BandagesAndRemedies",
  Containers = "Containers",
  RopesAndChains = "RopesAndChains",
  ThievesTools = "ThievesTools",
  ToolsOfTheTrade = "ToolsOfTheTrade",
  OrienteeringAids = "OrienteeringAids",
  Jewelry = "Jewelry",
  GemsAndPreciousStones = "GemsAndPreciousStones",
  Stationary = "Stationary",
  Books = "Books",
  MagicalArtifacts = "MagicalArtifacts",
  Alchemicae = "Alchemicae",
  Poisons = "Poisons",
  HealingHerbs = "HealingHerbs",
  MusicalInstruments = "MusicalInstruments",
  LuxuryGoods = "LuxuryGoods",
  Animals = "Animals",
  AnimalCare = "AnimalCare",
  Vehicles = "Vehicles",
  AusruestungDerGeweihtenschaft = "AusruestungDerGeweihtenschaft",
  CeremonialItems = "CeremonialItems",
  Liebesspielzeug = "Liebesspielzeug",
}

/**
 * Note text.
 * @markdown
 * @minLength 1
 */
type Note = string

/**
 * Special rules text.
 * @markdown
 * @minLength 1
 */
type Rules = string

/**
 * The weapon/armor advantage text.
 * @markdown
 * @minLength 1
 */
type Advantage = string

/**
 * The weapon/armor disadvantage text.
 * @markdown
 * @minLength 1
 */
type Disadvantage = string

export const validateSchema = validateSchemaCreator<Item>(import.meta.url)

File: Sources/OptolithDatabaseSchema/Types/equipment/item/_Item.swift
import FileDB

export const DefaultItem = (singularName: string, pluralName?: string) =>
  Entity(import.meta.url, {
    name: singularName,
    namePlural: pluralName ?? `${singularName}s`,
    type: () =>
      Object({
        cost: Required({
          comment: "The cost in silverthalers.",
          type: IncludeIdentifier(Cost),
        }),
        weight: Required({
          comment: "The weight in kg.",
          type: IncludeIdentifier(Weight),
        }),
        complexity: Optional({
          comment: "The complexity of crafting the item.",
          type: IncludeIdentifier(Complexity),
        }),
        structure_points: Required({
          comment:
            "The structure points of the item. Use an array if the item consists of multiple components that have individual structure points.",
          type: IncludeIdentifier(StructurePoints),
        }),
        combat_use: Optional({
          comment:
            "The item can also be used either as an improvised weapon or as an armor, although this is not the primary use case of the item.",
          type: IncludeIdentifier(CombatUse),
        }),

    /// The publications where you can find the entry.
    let src: PublicationRefs
        translations: DefaultItemTranslations(singularName),
      }),
    displayName: {},
  })

export const DefaultItemTranslations = (singularName: string) =>
  NestedLocaleMap(
    Required,
    `${singularName}Translation`,
    Object({

  /// The item’s name.
  let name: String({ minLength: 1 })
  /// An auxiliary name or label of the item, if available.
  let secondary_name: String({ minLength: 1 })?
  /// Note text.
  let note: String({ minLength: 1, isMarkdown: true })?
  /// Special rules text.
  let rules: String({ minLength: 1, isMarkdown: true })?
      errata: Optional({
        type: IncludeIdentifier(Errata),
      }),
    })
  )

/// The item can also be used either as an improvised weapon or as an armor, although this is not the primary use case of the item.
@ModelEnum
public enum CombatUse {
    case Weapon(IncludeIdentifier(ImprovisedWeapon))
    case Armor(IncludeIdentifier(SecondaryArmor))
}

export const StructurePoints = TypeAlias(import.meta.url, {
  name: "StructurePoints",
  comment:
    "The structure points of the item. Use an array if the item consists of multiple components that have individual structure points.",
  type: () => Array(IncludeIdentifier(StructurePointsComponent), { minItems: 1 }),
})

@Embedded
public struct StructurePointsComponent {

  /// The structure points.
  let points: Integer({ minimum: 1 })
  }

/// The cost in silverthalers.
@ModelEnum
public enum Cost {
    case Free
    case Various
    case Invaluable
    case Fixed(IncludeIdentifier(FixedCost))
    case Range(IncludeIdentifier(CostRange))
}

const wrap_in_text = Required({
  comment: "The cost get wrapped by this text using a placeholder in the text.",
  type: String({ minLength: 1, pattern: /\{0\}/ }),
})

@Embedded
public struct FixedCost {

  /// The cost in silverthalers.
  let value: Float({ minimum: { value: 0, isExclusive: true } })
      translations: NestedLocaleMap(
        Optional,
        "FixedCostTranslation",
        Object({
          wrap_in_text,
        })
      ),
  }

@Embedded
public struct CostRange {

  /// The lower bound of the cost in silverthalers.
  let from: Float({ minimum: { value: 0, isExclusive: true } })

  /// The upper bound of the cost in silverthalers.
  let to: Float({ minimum: { value: 0, isExclusive: true } })
      translations: NestedLocaleMap(
        Optional,
        "CostRangeTranslation",
        Object({
          wrap_in_text,
        })
      ),
  }

export const Weight = TypeAlias(import.meta.url, {
  name: "Weight",
  comment: "The weight in kg.",
  type: () => Float({ minimum: { value: 0, isExclusive: true } }),
})

/// The complexity of crafting the item.
@ModelEnum
public enum Complexity {
    case Primitive
    case Simple
    case Complex(IncludeIdentifier(ComplexComplexity))
}

@Embedded
public struct ComplexComplexity {

  /// The AP value for the trade secret.
  let ap_value: Integer({ minimum: 1 })
  }
